### PR TITLE
Fix piston retraction checking one block further than it needs to

### PR DESCRIPTION
**Pull request by jacob1**
Found this issue on my server, some players have this crazy ~~lag machine~~ mob pushing piston chain, which is nearby an admin claim. One of the chains is right near the border. A zero-ticked sticky piston is able to push a block right up to the border between the admin claim and the user claim, but on retract it gets stuck. This is because it's checking one block further than it needs to on retract, it sees that the air block past the block to retract is in another claim, so it cancels the event.

### CODE REVIEW

**Comment by Jikoo** [comment]
That doesn't make sense to me as described, could you please post an image? Also need to know what piston mode you're talking about.
Retraction definitely does also need to add direction - you have to remember that pistons can affect more than just the line in front of them, honey/slime can cause them to reach beside and behind themselves.

**Comment by Jikoo** [comment]
Oh I see. The direction of the piston is incorrect for retract. Your fix isn't correct, but it's on the right track.

/e Original snippet was wrong. What you should be doing is correcting the direction for retraction, i.e.
```Java
if (isRetract)
    direction = direction.getOppositeFace();
```

**Comment by jacob1** [comment]
Thought about it some more, you are correct. I didn't account for slime blocks.

Also I cat get a screenshot now, but to kind of describe the situation with ascii, the sticky piston will alternate between these two states:

| | |B|S
| |B| |S
where S is a sticky piston facing west, and B is a block. The boundary between the two claims is to the left of B in the bottom diagram. So alternating between these two states should never cause it to fail.
While extending, it does need to add one in the direction being pushed. While retracting, I had originally assumed nothing needed to be done, since we already know there's space to pull the blocks back in. But I didn't account for slime blocks. I think it would be possible to pull blocks into a claim with slime blocks, at the corner of the claim.

So yeah, your fix looks exactly correct. I can't test it or change it until tomorrow, feel free to commit it yourself if you want.

**Comment by Jikoo** [comment]
Opened jacob1/GriefPrevention#1 - looks like you unchecked the option allowing collaborators to commit to the PR's branch and I don't want to open a second PR for the same thing.

**Comment by jacob1** [comment]
Weird, I wouldn't have intentionally selected that option, I must have not noticed it was deselected. (Edit: you need "Maintainer" status to modify the PR according to the checkbox)

Anyway, I merged your PR to my PR. Haven't tested it, but it looks very correct to me.

**Comment by RoboMWM** [comment]
Not sure if I missed this notification or what but here I am now.

File: src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java [hunk 1]
```diff
@@ -570,7 +575,7 @@ private void onPistonEvent(BlockPistonEvent event, List<Block> blocks)
         }
 
         // Pushing down or pulling up is safe if all blocks are in line with the piston.
-        if (minX == maxX && minZ == maxZ && direction == (event instanceof BlockPistonExtendEvent ? BlockFace.DOWN : BlockFace.UP)) return;
+        if (minX == maxX && minZ == maxZ && direction == (isRetract ? BlockFace.UP : BlockFace.DOWN)) return;
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java:591]: idk if it was always like this but I rather have this notation expanded for readability - but I'll let it slide for now.

**Review by RoboMWM — APPROVED** [review]

**Comment by LogGits** [comment]
So this fixes the lag machines? I player on my server made one of these farms https://www.youtube.com/watch?v=OhlJb-_Qa1A and the tps went down to 3, with the timings noting that pistion extend event was the culprit. I'll compile the master branch and test it out when i can :)

**Comment by Jikoo** [comment]
This PR specifically is a fix for an issue with the piston changes from #933. Latest build of master is available [here](https://ci.appveyor.com/project/RoboMWM39862/griefprevention/build/artifacts). New config setting from 933 is documented [here](https://github.com/TechFortress/GriefPrevention/wiki/Setup-and-Configuration#pistonmovement).

The changes do improve GP's handling of pistons, but there's only so much that can be done while still actually protecting everything. `CLAIMS_ONLY` and `EVERYWHERE_SIMPLE` are certainly a lot more performant than the old counterparts, while `EVERYWHERE` will see only minor improvements when pistons push across claims.
One person has reported seeing functionality issues with `EVERYWHERE_SIMPLE` and heavily claimed/subclaimed areas, but I haven't been able to get results in my own testing/limited live runs. I assume the issues had to do with the (numerous, oops. I really thought my testing was pretty rigorous, egg on my face.) issues that have now been patched regarding the piston changes.

**Pull request merged**